benchmark/displace.py

```diff
@@ -0,0 +1,69 @@
+"""
+Eager implementation of displacement generation benchmark
+
+Created on Wed Jun 10 12:44:33 2020
+
+@author: Henry Liu
+"""
+import numpy as np
+import scipy as sc
+import tensorflow as tf
+
+import utils
+
+
+def gen_displace_scipy(N):
+    a = np.diag(np.sqrt(np.arange(1, N, dtype=np.complex64)), k=1)
+    a_dag = np.conjugate(a.transpose())
+    vexpm = np.vectorize(sc.linalg.expm, signature="(n,n)->(n,n)")
+
+    def displace(alphas):
+        alphas = np.expand_dims(np.expand_dims(alphas, -1), -1)
+        return vexpm(alphas * a_dag - np.conj(alphas) * a)
+
+    return displace
+
+
+def gen_displace(N):
+    a = utils.creation(N)
+    a_dag = tf.linalg.adjoint(a)
+
+    def displace(alphas):
+        # wrap for broadcasting [num, 1, 1]
+        alphas = tf.expand_dims(tf.expand_dims(alphas, -1), -1)
+        return tf.linalg.expm(alphas * a_dag - tf.math.conj(alphas) * a)
+
+    return displace
+
+
+def gen_displace_BCH(N):
+    a = tf.cast(utils.creation(N), dtype=tf.complex128)
+    a_dag = tf.linalg.adjoint(a)
+
+    # Convert to pos, mom basis
+    q = a_dag + a
+    p = 1j * (a_dag - a)
+
```

Comment:
Can you use a different definition of q and p (with a `sqrt(2)`). The problem with this definition is that their commutator is `2i` instead of `i` as it should be for position and momentum. This is mostly a notation issue, and as long as everything is consistent it shouldn't matter how they are defined. But I was using a different definition in the rest of the project, and I am afraid that this will lead to different factor of 2 mistakes in the future. 
```
        q = (a_dag + a) / sqrt(2)
        p = 1j*(a_dag - a) / sqrt(2)
```

Reply:
I double checked the derivation here and I think this is slightly tricky because it changes the form from exp(X) to exp(aX) = exp(X)^a. We have diagonal matrices X which means we can do this elementwise, but it adds a 3N pow operations and maybe additional numerical instability.

I think in this case it’s better to avoid the additional computation and rename the operators/add a note?

Reply:
Oops, ignore that - silly mistake, forgot that I can just include it inside of the exp call, this is all fine.

There is still some extra multiplications that we add, since we have to multiply all of the alphas by sqrt2. This is not really expensive given the other operations but I'm thinking it might still be better to define new variables `qsqrt2 = a_dag + a` or something.